service/builder/pipelines_builder.go

```diff
@@ -203,36 +205,104 @@ func (pb *PipelinesBuilder) getBuiltExportersByNames(exporterNames []string) []*
 	builtExporters := pb.getBuiltExportersByNames(exporterNames)
 
 	// Optimize for the case when there is only one exporter, no need to create junction point.
 	if len(builtExporters) == 1 {
 		return builtExporters[0].te
 	}
 
-	var exporters []consumer.TraceConsumer
+	var exporters []consumer.BaseTraceConsumer
 	for _, builtExp := range builtExporters {
 		exporters = append(exporters, builtExp.te)
 	}
 
 	// Create a junction point that fans out to all exporters.
-	return processor.NewTraceFanOutConnector(exporters)
+	return processor.CreateTraceFanOutConnector(exporters)
 }
 
-func (pb *PipelinesBuilder) buildFanoutExportersMetricsConsumer(exporterNames []string) consumer.MetricsConsumer {
+func (pb *PipelinesBuilder) buildFanoutExportersMetricsConsumer(exporterNames []string) consumer.BaseMetricsConsumer {
 	builtExporters := pb.getBuiltExportersByNames(exporterNames)
 
 	// Optimize for the case when there is only one exporter, no need to create junction point.
 	if len(builtExporters) == 1 {
 		return builtExporters[0].me
 	}
 
-	var exporters []consumer.MetricsConsumer
+	var exporters []consumer.BaseMetricsConsumer
 	for _, builtExp := range builtExporters {
 		exporters = append(exporters, builtExp.me)
 	}
 
 	// Create a junction point that fans out to all exporters.
-	return processor.NewMetricsFanOutConnector(exporters)
+	return processor.CreateMetricsFanOutConnector(exporters)
+}
+
+// createTraceProcessor creates trace processor based on type of the current processor
+//  and type of the downstream consumer.
+func createTraceProcessor(
+	factory processor.BaseFactory,
+	logger *zap.Logger,
+	cfg configmodels.Processor,
+	nextConsumer consumer.BaseTraceConsumer,
+) (processor.BaseTraceProcessor, error) {
+	if factoryV2, ok := factory.(processor.FactoryV2); ok {
+		creationParams := processor.CreationParams{Logger: logger}
+		ctx := context.Background()
+
+		// If both processor and consumer are of the new type (can manipulate on internal data structure),
+		// use FactoryV2.CreateTraceProcessor.
+		if nextConsumerV2, ok := nextConsumer.(consumer.TraceConsumerV2); ok {
+			return factoryV2.CreateTraceProcessor(ctx, creationParams, nextConsumerV2, cfg)
+		}
+
+		// If processor is of the new type, but downstream consumer is of the old type,
+		// use internalToOCTraceConverter compatibility shim.
+		traceConverter := consumer.NewInternalToOCTraceConverter(nextConsumer.(consumer.TraceConsumer))
+		return factoryV2.CreateTraceProcessor(ctx, creationParams, traceConverter, cfg)
+	}
+
+	// If both processor and consumer are of the old type (can manipulate on OC traces only),
+	// use Factory.CreateTraceProcessor.
+	if nextConsumer, ok := nextConsumer.(consumer.TraceConsumer); ok {
+		return factory.(processor.Factory).CreateTraceProcessor(logger, nextConsumer, cfg)
+	}
+
+	// Old type processor and a new type consumer usecase is not supported.
+	return nil, errors.New("OC Traces -> internal data format translation is not supported")
```

Comment:
Can you add a TODO to also support this case? We now have Oc->internal translation so this should not be difficult to implement. This will allow us to mix old and new style components in the pipeline.
We can implement it in a separate PR.

Reply:
Yes, makes sense. Updated